File: solutions/node/src/charging/domain/ChargingSession.ts
import type { ChargingEvent } from './events.ts';
import { chargerIdOf, parseConnectorId, type ConnectorId } from './Connector.ts';
import { parseIdTag, type IdTag } from './IdTag.ts';
import { wattHours, type WattHours } from './Energy.ts';
import { InvalidStateError } from './errors.ts';

export const SessionStatus = {
  Idle: 'Idle',
  Charging: 'Charging',
  Completed: 'Completed',
  Faulted: 'Faulted',
} as const;
export type SessionStatus = (typeof SessionStatus)[keyof typeof SessionStatus];

export interface StartInput {
  sessionId: string;
  idTag: IdTag;
  authorized: boolean; // decided by the AuthorizationService port, passed in (aggregate does no I/O)
  meterStartWh: WattHours;
  at: string;
}

// Aggregate root of the Charging context.
// Invariant: one connector has at most one session in status Charging.
// The aggregate only records events (R8); it never touches a repository, bus or clock.
export class ChargingSession {
  readonly connectorId: ConnectorId;
  private _status: SessionStatus;
  private _sessionId?: string;
  private _idTag?: IdTag;
  private _meterStartWh: WattHours;
  private _lastMeterWh: WattHours;
  private _startedAt?: string;
  private events: ChargingEvent[] = [];

  private constructor(
    connectorId: ConnectorId,
    status: SessionStatus,
    sessionId?: string,
    idTag?: IdTag,
    meterStartWh: WattHours = 0,
    lastMeterWh: WattHours = 0,
    startedAt?: string,
  ) {
    this.connectorId = connectorId;
    this._status = status;
    this._sessionId = sessionId;
    this._idTag = idTag;
    this._meterStartWh = meterStartWh;
    this._lastMeterWh = lastMeterWh;
    this._startedAt = startedAt;
  }

  // A fresh, not-yet-started session for a connector.
  static idle(connectorId: string): ChargingSession {
    return new ChargingSession(parseConnectorId(connectorId), SessionStatus.Idle);
  }

  // Used by repositories to rebuild state (Day 6). No events are recorded.
  static rehydrate(state: {
    connectorId: string;
    status: SessionStatus;
    sessionId?: string;
    idTag?: string;
    meterStartWh: number;
    lastMeterWh: number;
    startedAt?: string;
  }): ChargingSession {
    return new ChargingSession(
      state.connectorId,
      state.status,
      state.sessionId,
      state.idTag,
      state.meterStartWh,
      state.lastMeterWh,
      state.startedAt,
    );
  }

  get status(): SessionStatus {
    return this._status;
  }
  get sessionId(): string | undefined {
    return this._sessionId;
  }
  get idTag(): IdTag | undefined {
    return this._idTag;
  }
  get meterStartWh(): WattHours {
    return this._meterStartWh;
  }
  get lastMeterWh(): WattHours {
    return this._lastMeterWh;
  }
  get startedAt(): string | undefined {
    return this._startedAt;
  }

  // R1 + R2
  start(input: StartInput): void {
    const idTag = parseIdTag(input.idTag);
    if (this._status === SessionStatus.Charging) {
      // R1: the original session stays untouched; we only record the rejection.
      this.record({
        type: 'ChargingStartRejected',
        connectorId: this.connectorId,
        idTag,
        reason: 'ConnectorOccupied',
        occurredAt: input.at,
      });
      return;
    }
    if (!input.authorized) {
      // R2
      this.record({
        type: 'ChargingStartRejected',
        connectorId: this.connectorId,
        idTag,
        reason: 'Unauthorized',
        occurredAt: input.at,
      });
      return;
    }
    this._sessionId = input.sessionId;
    this._idTag = idTag;
    this._meterStartWh = wattHours(input.meterStartWh);
    this._lastMeterWh = this._meterStartWh;
    this._startedAt = input.at;
    this._status = SessionStatus.Charging;
    this.record({
      type: 'ChargingStarted',
      sessionId: input.sessionId,
      connectorId: this.connectorId,
      idTag,
      meterStartWh: this._meterStartWh,
      occurredAt: input.at,
    });
  }

  // R3: meter readings must be monotonic (a register never goes backwards).
  reportMeter(meterWh: WattHours, at: string): void {
    const value = wattHours(meterWh);
    const base = {
      sessionId: this._sessionId ?? '',
      connectorId: this.connectorId,
      meterWh: value,
      lastMeterWh: this._lastMeterWh,
      occurredAt: at,
    };
    if (this._status !== SessionStatus.Charging) {
      this.record({ type: 'MeterValueRejected', ...base, reason: 'NotCharging' });
      return;
    }
    if (value < this._lastMeterWh) {
      this.record({ type: 'MeterValueRejected', ...base, reason: 'NotMonotonic' });
      return;
    }
    this._lastMeterWh = value;
    this.record({
      type: 'EnergyMetered',
      sessionId: this._sessionId!,
      connectorId: this.connectorId,
      meterWh: value,
      occurredAt: at,
    });
  }

  // R4: only a Charging session can be stopped; energy = last meter - start meter.
  stop(stopReason: string, at: string): void {
    if (this._status !== SessionStatus.Charging) throw new InvalidStateError('stop', this._status);
    this._status = SessionStatus.Completed;
    this.record({
      type: 'ChargingCompleted',
      sessionId: this._sessionId!,
      connectorId: this.connectorId,
      energyWh: this._lastMeterWh - this._meterStartWh,
      startedAt: this._startedAt!,
      endedAt: at,
      stopReason,
      occurredAt: at,
    });
  }

  // The charger declared it cannot deliver energy per contract. A charging session is interrupted.
  reportFault(faultCode: string, at: string): void {
    const stillEnergized = this._status === SessionStatus.Charging;
    this._status = SessionStatus.Faulted;
    this.record({
      type: 'ChargerFaulted',
      chargerId: chargerIdOf(this.connectorId),
      connectorId: this.connectorId,
      faultCode,
      stillEnergized,
      occurredAt: at,
    });
  }

  // Application service calls this once, after the command succeeded, to hand events to the Outbox.
  pullEvents(): ChargingEvent[] {
    const pulled = this.events;
    this.events = [];
    return pulled;
  }

  private record(event: ChargingEvent): void {
    this.events.push(event);
  }
}

File: solutions/python/src/charging/domain/charging_session.py
"""Aggregate root of the Charging context.

Invariant: one connector has at most one session in status Charging.
The aggregate only records events (R8); it never touches a repository, bus or clock.
"""
from enum import Enum

from charging.domain.connector import charger_id_of, parse_connector_id
from charging.domain.energy import watt_hours
from charging.domain.errors import InvalidStateError
from charging.domain.events import (
    ChargerFaulted,
    ChargingCompleted,
    ChargingEvent,
    ChargingStarted,
    ChargingStartRejected,
    EnergyMetered,
    MeterValueRejected,
)
from charging.domain.id_tag import parse_id_tag


class SessionStatus(str, Enum):
    IDLE = "Idle"
    CHARGING = "Charging"
    COMPLETED = "Completed"
    FAULTED = "Faulted"


class ChargingSession:
    def __init__(self, connector_id: str, status: SessionStatus = SessionStatus.IDLE) -> None:
        self.connector_id = parse_connector_id(connector_id)
        self.status = status
        self.session_id: str | None = None
        self.id_tag: str | None = None
        self.meter_start_wh: int = 0
        self.last_meter_wh: int = 0
        self.started_at: str | None = None
        self._events: list[ChargingEvent] = []

    @classmethod
    def idle(cls, connector_id: str) -> "ChargingSession":
        """A fresh, not-yet-started session for a connector."""
        return cls(connector_id)

    # R1 + R2
    def start(self, *, session_id: str, id_tag: str, authorized: bool, meter_start_wh: int, at: str) -> None:
        tag = parse_id_tag(id_tag)
        if self.status is SessionStatus.CHARGING:
            # R1: the original session stays untouched; we only record the rejection.
            self._record(ChargingStartRejected(occurred_at=at, connector_id=self.connector_id, id_tag=tag, reason="ConnectorOccupied"))
            return
        if not authorized:  # R2
            self._record(ChargingStartRejected(occurred_at=at, connector_id=self.connector_id, id_tag=tag, reason="Unauthorized"))
            return
        self.session_id = session_id
        self.id_tag = tag
        self.meter_start_wh = watt_hours(meter_start_wh)
        self.last_meter_wh = self.meter_start_wh
        self.started_at = at
        self.status = SessionStatus.CHARGING
        self._record(
            ChargingStarted(occurred_at=at, session_id=session_id, connector_id=self.connector_id, id_tag=tag, meter_start_wh=self.meter_start_wh)
        )

    # R3: meter readings must be monotonic (a register never goes backwards).
    def report_meter(self, meter_wh: int, at: str) -> None:
        value = watt_hours(meter_wh)
        rejected = dict(occurred_at=at, session_id=self.session_id or "", connector_id=self.connector_id, meter_wh=value, last_meter_wh=self.last_meter_wh)
        if self.status is not SessionStatus.CHARGING:
            self._record(MeterValueRejected(reason="NotCharging", **rejected))
            return
        if value < self.last_meter_wh:
            self._record(MeterValueRejected(reason="NotMonotonic", **rejected))
            return
        self.last_meter_wh = value
        self._record(EnergyMetered(occurred_at=at, session_id=self.session_id or "", connector_id=self.connector_id, meter_wh=value))

    # R4: only a Charging session can be stopped; energy = last meter - start meter.
    def stop(self, stop_reason: str, at: str) -> None:
        if self.status is not SessionStatus.CHARGING:
            raise InvalidStateError("stop", self.status.value)
        self.status = SessionStatus.COMPLETED
        self._record(
            ChargingCompleted(
                occurred_at=at,
                session_id=self.session_id or "",
                connector_id=self.connector_id,
                energy_wh=self.last_meter_wh - self.meter_start_wh,
                started_at=self.started_at or "",
                ended_at=at,
                stop_reason=stop_reason,
            )
        )

    def report_fault(self, fault_code: str, at: str) -> None:
        """The charger declared it cannot deliver energy per contract. A charging session is interrupted."""
        still_energized = self.status is SessionStatus.CHARGING
        self.status = SessionStatus.FAULTED
        self._record(
            ChargerFaulted(
                occurred_at=at,
                charger_id=charger_id_of(self.connector_id),
                connector_id=self.connector_id,
                fault_code=fault_code,
                still_energized=still_energized,
            )
        )

    def pull_events(self) -> list[ChargingEvent]:
        """Application service calls this once, after the command succeeded, to hand events to the Outbox."""
        pulled, self._events = self._events, []
        return pulled

    def _record(self, event: ChargingEvent) -> None:
        self._events.append(event)
